protocol/swap/swap_manager.go

```diff
@@ -140,40 +147,86 @@ func (m *manager) GetPastSwap(id types.Hash) (*Info, error) {
 	return s, nil
 }
 
-// GetOngoingSwap returns the ongoing swap's *Info, if there is one.
-func (m *manager) GetOngoingSwap(id types.Hash) (Info, error) {
+// GetOngoingSwap returns the ongoing swap's *Info, if there is one. The
+// returned Info structure of an active swap can be modified as the swap's state
+// changes and should only be read or written by a single go process.
+func (m *manager) GetOngoingSwap(offerID types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	s, has := m.ongoing[id]
+
+	s, has := m.ongoing[offerID]
 	if !has {
-		return Info{}, errNoSwapWithID
+		return nil, errNoSwapWithOfferID
 	}
 
-	return *s, nil
+	return s, nil
 }
 
-// GetOngoingSwaps returns all ongoing swaps.
-func (m *manager) GetOngoingSwaps() ([]*Info, error) {
+// GetOngoingSwapSnapshot returns a copy of the ongoing swap's Info, if the
+// offerID has an ongoing swap.
+func (m *manager) GetOngoingSwapSnapshot(offerID types.Hash) (*Info, error) {
 	m.RLock()
 	defer m.RUnlock()
-	swaps := make([]*Info, len(m.ongoing))
-	i := 0
+
+	s, has := m.ongoing[offerID]
+	if !has {
+		return nil, errNoSwapWithOfferID
+	}
+
+	// We have the read lock above when making this copy, but the swapState
+	// instances don't use that lock when they modify the same Info structure
+	// from a different go process. At some point we should update the swapState
+	// instances to make all changes to their ongoing Info structure via this
+	// manager.
```

Comment:
we should open an issue for this; ideally we don't need the snapshot functions and it's only copies that are ever returned, and the other processes modify the swap info (I think status is the only field modified after creation?) via this manager

Reply:
> I think status is the only field modified after creation?

`EndTime`, `LastStatusUpdateTime`, `Timeout1` and `Timeout2` are also modified after creation.

Reply:
> we should open an issue for this

The locking is now in this PR, so we don't need to open an issue. I removed the comment above. The locking in here is exclusively for the map maintained by the manager. The lock for the `Info` structure is inside the struct and used by member functions of the struct.